File: imageviewerlib/src/main/java/cn/yzapp/imageviewerlib/ImageViewerActivity.java
package cn.yzapp.imageviewerlib;

import android.animation.ValueAnimator;
import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.LinearInterpolator;
import android.widget.RelativeLayout;

import java.io.File;

import me.relex.circleindicator.CircleIndicator;
import uk.co.senab.photoview.PhotoViewAttacher;

/**
 * @author nestor
 *         email nestor@yzapp.cn
 */
public class ImageViewerActivity extends Activity {

    private ImageInfo mShowImage;
    private LocalBroadcastManager mLocalBroadcastManager;
    private SamplePagerAdapter mSamplePagerAdapter;
    private RelativeLayout root;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mLocalBroadcastManager = LocalBroadcastManager.getInstance(this);

        setContentView(R.layout.activity_image_viewer);
        root = (RelativeLayout) findViewById(R.id.root);

        mShowImage = getIntent().getParcelableExtra(ImageViewer.INTENT_IMAGE);

        setViewPage();

        setOpenAnimator();

    }

    private void setViewPage() {
        final HackyViewPager mViewPager = (HackyViewPager) findViewById(R.id.view_pager);

        mSamplePagerAdapter = new SamplePagerAdapter();
        mViewPager.setAdapter(mSamplePagerAdapter);

        setCircleIndicator(mViewPager);

        mViewPager.setCurrentItem(mShowImage.getIndex());
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                sendBroadcast(position);
                SmoothImageView smoothImageView = mSamplePagerAdapter.getPhotoView(position);
                if (smoothImageView != null) {
                    smoothImageView.setScale(1.0f);
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {
            }
        });
    }

    private void setCircleIndicator(HackyViewPager mViewPager) {
        CircleIndicator mIndicator = (CircleIndicator) findViewById(R.id.indicator);
        int indicatorBackgroundId = getIntent().getIntExtra(ImageViewer.CHOOSE_RES_IS, 0);
        int indicatorUnselectedBackgroundId = getIntent().getIntExtra(ImageViewer.UNCHOOSE_RES_IS, 0);
        mIndicator.configureIndicator(-1, -1, -1, 0, 0, indicatorBackgroundId, indicatorUnselectedBackgroundId);

        if(mShowImage.getImg().size() > 1){
            mIndicator.setVisibility(View.VISIBLE);
            mIndicator.setViewPager(mViewPager);
        }
    }

    private void setOpenAnimator() {
        // 进度动画
        ValueAnimator animator = ValueAnimator.ofInt(0, 0xFF);
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                int progress = (int) animation.getAnimatedValue();
                root.setBackgroundColor(progress * 0x1000000);
            }
        });
        // 设置插值器为均速滚动
        animator.setInterpolator(new LinearInterpolator());
        animator.setDuration(380);
        animator.start();
    }

    private void sendBroadcast(int position) {
        Intent intent = new Intent();
        intent.setAction(ImageViewer.BROADCAST_ACTION);
        intent.putExtra("position", position);
        mLocalBroadcastManager.sendBroadcast(intent);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (isFinishing()) {
            overridePendingTransition(0, 0);
        }
    }

    @Override
    protected void onDestroy() {
        Intent intent = new Intent();
        intent.setAction(ImageViewer.BROADCAST_ACTION);
        intent.putExtra("position", -1);
        intent.putExtra("onDestroy", true);
        mLocalBroadcastManager.sendBroadcast(intent);

        super.onDestroy();
    }

    private void setPhotoView(final SmoothImageView photoView, Object img) {
        if (img instanceof String) {
            ImageViewerConfig.getImageLoader().getImage(this, photoView, (String) img);
        }
        if (img instanceof Integer) {
            ImageViewerConfig.getImageLoader().getImage(this, photoView, (int) img);
        }
        if (img instanceof File) {
            ImageViewerConfig.getImageLoader().getImage(this, photoView, (File) img);
        }
        if (img instanceof Bitmap) {
            ImageViewerConfig.getImageLoader().getImage(this, photoView, (Bitmap) img);
        }

        photoView.setOnTransformListener(new SmoothImageView.TransformListener() {
            @Override
            public void onTransformComplete(int mode) {
                if (mode == 2) {
                    finish();
                }
            }
        });

        photoView.setOnPhotoTapListener(new PhotoViewAttacher.OnPhotoTapListener() {
            @Override
            public void onPhotoTap(View view, float v, float v2) {
                photoView.transformOut();
                setCloseAnimator();

            }

            @Override
            public void onOutsidePhotoTap() {
                photoView.transformOut();
                setCloseAnimator();
            }
        });
    }

    private void setCloseAnimator() {
        // 进度动画
        ValueAnimator animator = ValueAnimator.ofInt(0xFF, 0);
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                int progress = (int) animation.getAnimatedValue();
                root.setBackgroundColor(progress * 0x1000000);
            }
        });
        // 设置插值器为均速滚动
        animator.setInterpolator(new LinearInterpolator());
        animator.setDuration(380);
        animator.start();
    }

    class SamplePagerAdapter extends PagerAdapter {
        private SparseArray<SmoothImageView> photoViews;
        private boolean show;

        public SamplePagerAdapter() {
            photoViews = new SparseArray<>(mShowImage.getImg().size());
        }

        @Override
        public int getCount() {
            return mShowImage.getImg().size();
        }

        @Override
        public View instantiateItem(ViewGroup container, int position) {
            SmoothImageView photoView = new SmoothImageView(container.getContext());
            setImgSite(position, photoView);

            if (position == mShowImage.getIndex() && !show) {
                photoView.transformIn();
                show = true;
            }
            setPhotoView(photoView, mShowImage.getImg().get(position));

            container.addView(photoView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

            photoViews.put(position, photoView);
            return photoView;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView((View) object);
            photoViews.remove(position);
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }

        public SmoothImageView getPhotoView(int position) {
            return photoViews.get(position);
        }

    }

    /**
     * 设置图片起始位置
     */
    private void setImgSite(int position, SmoothImageView photoView) {
        int[] size = null;
        if (mShowImage.getSizes() != null && mShowImage.getSizes().size() > position) {
            size = mShowImage.getSizes().get(position);
        }

        if (size == null || size.length < 4) {
            photoView.setOriginalInfo(
                    0,
                    0,
                    Utils.getScreenWidth(ImageViewerActivity.this) / 2,
                    Utils.getScreenHeight(ImageViewerActivity.this) / 2);
        } else {
            photoView.setOriginalInfo(
                    size[0],
                    size[1],
                    size[2],
                    size[3]);
        }
    }
}

File: imageviewerlib/src/main/java/cn/yzapp/imageviewerlib/ImageViewerActivity.kt
package cn.yzapp.imageviewerlib

import android.animation.ValueAnimator
import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.os.Build
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.util.SparseArray
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import kotlinx.android.synthetic.main.activity_image_viewer.*

import java.io.File

import uk.co.senab.photoview.PhotoViewAttacher
import kotlin.properties.Delegates

/**
 * @author nestor
 * *         email nestor@yzapp.cn
 */
class ImageViewerActivity : Activity() {

    private var mShowImage: ImageInfo by Delegates.notNull()
    private var mLocalBroadcastManager: LocalBroadcastManager by Delegates.notNull()
    private val mSamplePagerAdapter: SamplePagerAdapter by lazy {
        SamplePagerAdapter()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_image_viewer)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
        }

        mLocalBroadcastManager = LocalBroadcastManager.getInstance(this)
        mShowImage = intent.getParcelableExtra(ImageViewer.INTENT_IMAGE)

        setViewPage()
        setOpenAnimator()
    }

    override fun onPause() {
        super.onPause()
        if (isFinishing) {
            overridePendingTransition(0, 0)
        }
    }

    override fun onDestroy() {
        val intent = Intent()
        intent.action = ImageViewer.BROADCAST_ACTION
        intent.putExtra("position", -1)
        intent.putExtra("onDestroy", true)
        mLocalBroadcastManager.sendBroadcast(intent)

        super.onDestroy()
    }

    private fun setViewPage() {
        view_pager.adapter = mSamplePagerAdapter

        setCircleIndicator(view_pager)

        view_pager.currentItem = mShowImage.index
        view_pager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

            }

            override fun onPageSelected(position: Int) {
                sendBroadcast(position)
                val smoothImageView = mSamplePagerAdapter.getPhotoView(position)
                smoothImageView?.scale = 1.0f
            }

            override fun onPageScrollStateChanged(state: Int) {
            }
        })
    }

    private fun setCircleIndicator(mViewPager: HackyViewPager) {
        val indicatorBackgroundId = intent.getIntExtra(ImageViewer.CHOOSE_RES_IS, 0)
        val indicatorUnselectedBackgroundId = intent.getIntExtra(ImageViewer.UNCHOOSE_RES_IS, 0)
        indicator.configureIndicator(-1, -1, -1, 0, 0, indicatorBackgroundId, indicatorUnselectedBackgroundId)

        if (mShowImage.img.size > 1) {
            indicator.visibility = View.VISIBLE
            indicator.setViewPager(mViewPager)
        }
    }

    private fun setOpenAnimator() {
        // 进度动画
        val animator = ValueAnimator.ofInt(0, 0xFF)
        animator.addUpdateListener { animation ->
            val progress = animation.animatedValue as Int
            root.setBackgroundColor(progress * 0x1000000)
        }
        // 设置插值器为均速滚动
        //animator.interpolator = LinearInterpolator()
        animator.duration = 380
        animator.start()
    }

    private fun setCloseAnimator() {
        // 进度动画
        val animator = ValueAnimator.ofInt(0xFF, 0)
        animator.addUpdateListener { animation ->
            val progress = animation.animatedValue as Int
            root.setBackgroundColor(progress * 0x1000000)
        }
        // 设置插值器为均速滚动
        //animator.interpolator = LinearInterpolator()
        animator.duration = 380
        animator.start()
    }

    private fun sendBroadcast(position: Int) {
        val intent = Intent()
        intent.action = ImageViewer.BROADCAST_ACTION
        intent.putExtra("position", position)
        mLocalBroadcastManager.sendBroadcast(intent)
    }

    internal inner class SamplePagerAdapter : PagerAdapter() {
        private val photoViews: SparseArray<SmoothImageView> = SparseArray(mShowImage.img.size)
        private var show: Boolean = false

        override fun getCount(): Int {
            return mShowImage.img.size
        }

        override fun instantiateItem(container: ViewGroup, position: Int): View {
            val photoView = SmoothImageView(container.context)
            setImgSite(position, photoView)

            if (position == mShowImage.index && !show) {
                photoView.transformIn()
                show = true
            }
            setPhotoView(photoView, mShowImage.img[position])

            container.addView(photoView, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

            photoViews.put(position, photoView)
            return photoView
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(`object` as View)
            photoViews.remove(position)
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view === `object`
        }

        fun getPhotoView(position: Int): SmoothImageView? {
            return photoViews.get(position)
        }

        /**
         * 设置图片起始位置
         */
        private fun setImgSite(position: Int, photoView: SmoothImageView) {
            var size: IntArray? = null
            if (mShowImage.sizes.size > position) {
                size = mShowImage.sizes[position]
            }

            if (size == null || size.size < 4) {
                photoView.setOriginalInfo(
                        0,
                        0,
                        Utils.getScreenWidth(this@ImageViewerActivity) / 2,
                        Utils.getScreenHeight(this@ImageViewerActivity) / 2)
            } else {
                photoView.setOriginalInfo(
                        size[0],
                        size[1],
                        size[2],
                        size[3])
            }
        }


        private fun setPhotoView(photoView: SmoothImageView, img: Any) {
            if (img is String) {
                ImageViewerConfig.imageLoader.getImage(this@ImageViewerActivity, photoView, img)
            }
            if (img is Int) {
                ImageViewerConfig.imageLoader.getImage(this@ImageViewerActivity, photoView, img)
            }
            if (img is File) {
                ImageViewerConfig.imageLoader.getImage(this@ImageViewerActivity, photoView, img)
            }
            if (img is Bitmap) {
                ImageViewerConfig.imageLoader.getImage(this@ImageViewerActivity, photoView, img)
            }

            photoView.setOnTransformListener(object : SmoothImageView.TransformListener {
                override fun onTransformComplete(mode: Int) {
                    if (mode == 2) {
                        finish()
                    }
                }
            })

            photoView.setOnPhotoTapListener(object : PhotoViewAttacher.OnPhotoTapListener {
                override fun onPhotoTap(view: View, v: Float, v2: Float) {
                    photoView.transformOut()
                    setCloseAnimator()

                }

                override fun onOutsidePhotoTap() {
                    photoView.transformOut()
                    setCloseAnimator()
                }
            })
        }

    }

}
